src/physics/systems.rs

```diff
@@ -182,3 +190,86 @@ pub fn sync_transform_system(
         }
     }
 }
+
+/// System responsible for removing joints, colliders, and bodies that have
+/// been removed from the scene
+pub fn destroy_body_and_collider_system(
+    mut commands: Commands,
+    mut bodies: ResMut<RigidBodySet>,
+    mut colliders: ResMut<ColliderSet>,
+    mut joints: ResMut<JointSet>,
+    mut entity_maps: ResMut<EntityMaps>,
+    collider_query: Query<(Entity, &ColliderHandleComponent)>,
+    joint_query: Query<(Entity, &JointHandleComponent)>,
+    body_query: Query<(Entity, &RigidBodyHandleComponent)>,
+) {
+    // Components removed before this system
+    let mut bodies_removed: HashSet<Entity> = body_query
+        .removed::<RigidBodyHandleComponent>()
+        .into_iter()
+        .map(|e| *e)
+        .collect();
+    let mut colliders_removed: HashSet<Entity> = collider_query
+        .removed::<ColliderHandleComponent>()
+        .into_iter()
+        .map(|e| *e)
+        .collect();
+    let mut joints_removed: HashSet<Entity> = joint_query
+        .removed::<JointHandleComponent>()
+        .into_iter()
+        .map(|e| *e)
+        .collect();
+
+    for entity in &bodies_removed.clone() {
+        if let Some(body) = entity_maps.bodies.get(&entity) {
+            bodies.remove(*body, &mut colliders, &mut joints);
+            entity_maps.bodies.remove(&entity);
+            bodies_removed.remove(entity);
+            // Removing a body also removes its colliders and joints. If they were
+            // not also removed then we must remove them here.
+            if !colliders_removed.contains(&entity) {
+                commands.remove_one::<ColliderHandleComponent>(*entity);
+            } else {
+                colliders_removed.remove(&entity);
+            }
+            if !joints_removed.contains(&entity) {
+                commands.remove_one::<JointHandleComponent>(*entity);
+            } else {
+                joints_removed.remove(&entity);
+            }
+        }
+    }
+    for entity in &colliders_removed.clone() {
+        if let Some(collider) = entity_maps.colliders.get(entity) {
+            colliders.remove(*collider, &mut bodies);
+            colliders_removed.remove(entity);
+            entity_maps.colliders.remove(&entity);
+        }
+    }
+    // FIXME: How should joints be removed?
+    // for entity in &joints_removed.clone() {
+    //     if let Some((joint, entity1, entity2)) = entity_maps.joints.get(entity) {
+    //         ??? joints.remove(*joint, &mut bodies);
```

Comment:
I don't know what to do about joints as `JointSet` does not have a `remove()` function.

Reply:
Something like this could have worked but `RigidBody.joint_graph_index` and `JointSet.remove_rigid_body()` are crate-private:
```rust
    for entity in &joints_removed.clone() {
        if let Some((joint, entity1, entity2)) = entity_maps.joints.get(entity) {
            if let Some(rbh1) = entity_maps.bodies.get(entity1) {
                if let Some(rb1) = bodies.get(*rbh1) {
                    joints.remove_rigid_body(rb1.joint_graph_index, &mut bodies);
                }
            }
            if let Some(rbh2) = entity_maps.bodies.get(entity2) {
                if let Some(rb2) = bodies.get(*rbh2) {
                    joints.remove_rigid_body(rb2.joint_graph_index, &mut bodies);
                }
            }
            joints_removed.remove(entity);
            entity_maps.joints.remove(entity);
        }
    }
```

Reply:
Joint removal will have to be implemented in Rapier first.

Reply:
Works now with rapier 0.3.1. :)